src/private.js

```diff
@@ -489,33 +489,47 @@ module.exports = (dht) => ({
       }
 
       // need more, query the network
-      const query = new Query(dht, key.buffer, (peer, cb) => {
-        waterfall([
-          (cb) => dht._findProvidersSingle(peer, key, cb),
-          (msg, cb) => {
-            const provs = msg.providerPeers
-            dht._log('(%s) found %s provider entries', dht.peerInfo.id.toB58String(), provs.length)
+      const paths = []
+      const query = new Query(dht, key.buffer, (pathIndex, numPaths) => {
+        const pathSize = utils.pathSize(out.length - n, pathIndex, numPaths)
+        const pathProviders = new LimitedPeerList(pathSize)
+        paths.push(pathProviders)
+
+        return (peer, cb) => {
+          waterfall([
+            (cb) => dht._findProvidersSingle(peer, key, cb),
+            (msg, cb) => {
+              const provs = msg.providerPeers
+              dht._log('(%s) found %s provider entries', dht.peerInfo.id.toB58String(), provs.length)
+
+              provs.forEach((prov) => {
+                pathProviders.push(dht.peerBook.put(prov))
+              })
 
-            provs.forEach((prov) => {
-              out.push(dht.peerBook.put(prov))
-            })
+              // hooray we have all that we want
+              if (pathProviders.length >= pathSize) {
+                return cb(null, {success: true})
+              }
 
-            // hooray we have all that we want
-            if (out.length >= n) {
-              return cb(null, {success: true})
+              // it looks like we want some more
+              cb(null, {
+                closerPeers: msg.closerPeers
+              })
             }
-
-            // it looks like we want some more
-            cb(null, {
-              closerPeers: msg.closerPeers
-            })
-          }
-        ], cb)
+          ], cb)
+        }
       })
 
       const peers = dht.routingTable.closestPeers(key.buffer, c.ALPHA)
 
       timeout((cb) => query.run(peers, cb), maxTimeout)((err) => {
+        // combine peers from each path
+        paths.forEach((path) => {
+          path.toArray().forEach((peer) => {
+            out.push(peer)
```

Comment:
Same as reported above

Reply:
Neither `out` nor `path` are arrays, so I don't see how the above solution applies here. What do you have in mind?

Reply:
I could just replace them with arrays, and either just allow extra elements in the return value, or slice off the extras.

Reply:
If I am understanding correctly, `out` is an array and you are transforming `path` to an array. So, I think we can use the same `concat` logic as I suggested above. Am I missing something?

Reply:
`out` isn't an array. Are you saying to make it one?

Reply:
Ok, since it is a LimitedPeerList, we can leave it this way